homeassistant/components/scheduler/schedule.py

```diff
@@ -0,0 +1,378 @@
+        # recurrence; since we've already established that there _is_ a recurrence,
+        # we're active:
+        last_start_dt = self.recurrence.before(now, inc=True)
+        if not last_start_dt:
+            return True
+
+        # Lastly, look at the recurrence just before right now and see
+        # if we're still inside it; if we are, we're active:
+        last_end_dt = last_start_dt + self.instance_duration
+        return last_start_dt <= now <= last_end_dt
+
+    @property
+    def expired(self) -> bool:
+        """Return whether the schedule has expired."""
+        now = dt_util.utcnow()
+
+        # If a recurrence exists and there's at least one instance of it in the future,
+        # we're not expired:
+        if self.recurrence and self.recurrence.after(now, inc=True):
+            return False
+
+        # If either the initial start datetime or the end datetime (if it exists) are in
+        # in the future, we're not expired:
+        if self.start_datetime >= now or (
+            self.end_datetime and self.end_datetime >= now
+        ):
+            return False
+
+        return True
+
+    @property
+    def state(self) -> str:
+        """Return the current state of the schdule."""
+        if self.expired:
+            return STATE_EXPIRED
+        if not self._is_on:
+            return STATE_OFF
+        return STATE_ON
+
+    @property
+    def is_on(self) -> bool:
+        """Return whether the schedule is enabled."""
+        return self._is_on
+
+    @callback
+    def _async_schedule(self) -> None:
+        """Schedule the next instance."""
+        if not self._is_on:
+            _LOGGER.warning("Schedule is turned off")
+            return
+
+        start_dt, end_dt = self._get_next_instance_datetimes()
+
+        if not start_dt:
+            _LOGGER.info("No more instances of schedule: %s", self)
+            return
+
+        instance = ScheduleInstance(self._hass, self.entity_id)
+
+        if self.instance_duration:
+            self.active_instance = instance
+
+        async def start(self, executed_at: datetime) -> None:
+            """Trigger."""
+            await self.async_trigger(instance)
+
+        async def stop(self, executed_at: datetime) -> None:
+            """Revert."""
+            await self.async_revert(instance)
+
+        self._async_trigger_listener = async_track_point_in_time(
+            self._hass, start, start_dt
+        )
+
+        if end_dt:
+            self._async_trigger_listener = async_track_point_in_time(
+                self._hass, stop, end_dt
+            )
+
+    @callback
+    def _get_next_instance_datetimes(
+        self,
+    ) -> Tuple[Optional[datetime], Optional[datetime]]:
+        """Get the next starting (and, optionally, ending) datetimes."""
+        if not self._initial_instance_scheduled:
+            self._initial_instance_scheduled = True
+            return (self.start_datetime, self.end_datetime)
+
+        if self.recurrence:
```

Comment:
Let's turn this into a guard class. `if not self.recurrence: return (None, None)`